dep0/lib/01_ast/include/dep0/ast/pretty_print.hpp

```diff
@@ -0,0 +1,385 @@
+
+template <Properties P>
+std::ostream& pretty_print(std::ostream&, typename expr_t<P>::var_t const&, std::size_t indent = 0ul);
+
+template <Properties P>
+std::ostream& pretty_print(std::ostream&, typename expr_t<P>::app_t const&, std::size_t indent = 0ul);
+
+template <Properties P>
+std::ostream& pretty_print(std::ostream&, typename expr_t<P>::abs_t const&, std::size_t indent = 0ul);
+
+// implementations
+
+inline std::ostream& new_line(std::ostream& os, std::size_t indent)
+{
+    os << std::endl;
+    while (indent--)
+        os << "    ";
+    return os;
+}
+
+template <Properties P>
+std::ostream& pretty_print(std::ostream& os, module_t<P> const& module, std::size_t const indent)
+{
+    bool first = true;
+    for (auto const& type_def: module.type_defs)
+        pretty_print(std::exchange(first, false) ? os : new_line(os, indent), type_def, indent);
+    for (auto const& func_def: module.func_defs)
+        pretty_print(std::exchange(first, false) ? os : new_line(os, indent), func_def, indent);
+    return os;
+}
+
+template <Properties P>
+std::ostream& pretty_print(std::ostream& os, type_def_t<P> const& type_def, std::size_t)
+{
+    match(
+        type_def.value,
+        [&] (typename type_def_t<P>::integer_t const& x)
+        {
+            os << (x.sign == sign_t::signed_v ? "signed" : "unsigned") << ' ';
+            os << (
+                x.width == width_t::_8 ? "8" :
+                x.width == width_t::_16 ? "16" :
+                x.width == width_t::_32 ? "32" :
+                "64") << " bit integer from ";
+            if (x.sign == sign_t::signed_v)
+            {
+                if (x.max_abs_value)
+                    os << '-' << *x.max_abs_value << " to " << *x.max_abs_value;
+                else
+                    os << "... to ...";
+            }
+            else
+                os << "0 to " << (x.max_abs_value ? x.max_abs_value->view() : "...");
+        });
+    return os;
+}
+
+template <Properties P>
+std::ostream& pretty_print(std::ostream& os, func_def_t<P> const& func_def, std::size_t const indent)
+{
+    os << "auto " << func_def.name;
+    if (func_def.value.args.empty())
+        os << "()";
+    else
+    {
+        os << '(';
+        for (bool first = true; auto const& arg: func_def.value.args)
+        {
+            new_line(std::exchange(first, false) ? os : os << ',', indent + 1ul);
+            pretty_print(os, arg.sort, indent + 1ul);
+            pretty_print(os << ' ', arg.name);
+        }
+        new_line(os, indent) << ')';
+    }
+    pretty_print(os << " -> ", func_def.value.ret_type, indent);
+    new_line(os, indent);
+    pretty_print(os, func_def.value.body, indent);
+    return os;
+}
+
+template <Properties P>
+std::ostream& pretty_print(std::ostream& os, sort_t<P> const& x, std::size_t const indent)
+{
+    match(
+        x,
+        [&] (typename_t const&) { os << "typename"; },
+        [&] (type_t<P> const& t) { pretty_print(os, t, indent); });
```

Comment:
could add overload for `typename_t` and this would become
```suggestion
    match(x, [&] (auto const& t) { pretty_print(os, t, indent); });
```